File: src/main/java/uk/gov/justice/hmpps/casenotes/services/EventListener.java
package uk.gov.justice.hmpps.casenotes.services;

import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.jms.annotation.JmsListener;
import org.springframework.stereotype.Service;
import uk.gov.justice.hmpps.casenotes.dto.OffenderEvent;

import java.io.IOException;
import java.util.Map;

@Service
@ConditionalOnProperty(name = "sqs.provider")
@Slf4j
@AllArgsConstructor
public class EventListener {

    private final ObjectMapper objectMapper;
    private final MergeOffenderService mergeOffenderService;

    @JmsListener(destination = "${sqs.queue.name}")
    public void eventListener(final String requestJson) {
        final var event = getOffenderEvent(requestJson);
        if (event != null) {
            mergeOffenderService.checkForMerge(event);
        }
    }

    private OffenderEvent getOffenderEvent(final String requestJson) {
        OffenderEvent event = null;
        try {
            final Map<String, String> message = objectMapper.readValue(requestJson, Map.class);
            if (message != null && message.get("Message") != null) {
                event = objectMapper.readValue(message.get("Message"), OffenderEvent.class);
            }
        } catch (IOException e) {
            log.error("Failed to Parse Message {}", requestJson);
        }
        return event;
    }


}

File: src/main/java/uk/gov/justice/hmpps/casenotes/services/EventListener.kt
package uk.gov.justice.hmpps.casenotes.services

import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import io.awspring.cloud.sqs.annotation.SqsListener
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

@Service
class EventListener(
  private val mergeOffenderService: MergeOffenderService,
  private val objectMapper: ObjectMapper,
) {
  companion object {
    private val log = LoggerFactory.getLogger(this::class.java)
  }

  @SqsListener("event", factory = "hmppsQueueContainerFactoryProxy")
  fun handleEvents(requestJson: String) {
    val (message, messageAttributes) = objectMapper.readValue<Message>(requestJson)
    val bookingId = objectMapper.readValue<EventMessage>(message).bookingId

    val eventType = messageAttributes.eventType.value
    log.info("Processing message of type {}", eventType)

    when (eventType) {
      "BOOKING_NUMBER-CHANGED" -> mergeOffenderService.checkAndMerge(bookingId)
    }
  }
}

data class Attribute(@JsonProperty("Type") val type: String, @JsonProperty("Value") val value: String)
data class MessageAttributes(val eventType: Attribute)
data class EventMessage(val offenderIdDisplay: String?, val bookingId: Long)
data class Message(
  @JsonProperty("Message") val message: String,
  @JsonProperty("MessageAttributes") val messageAttributes: MessageAttributes,
)
